tests/api/hello_with_curl/CMakeLists.txt

```diff
@@ -0,0 +1,164 @@
+cmake_minimum_required(VERSION 3.15)
+
+project(hello VERSION 1.0.0.0)
+
+if(NOT DEFINED CMAKE_JS_VERSION)
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/node_modules/cmake-js/share/cmake")
+endif()
+
+# Get CURL::libcurl
+find_package(CURL REQUIRED)
+
+if(hello_IS_INSTALLABLE)
+  include(GNUInstallDirs)
+  # cURL relocate...
+  if(NOT EXISTS "${CMAKE_CURRENT_BINARY_DIR}/include/curl")
+    # Set up cURL for transport
+    file(COPY "${CURL_INCLUDE_DIR}/curl" DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/include")
+  endif()
+  set (CURL_INCLUDE_DIR
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/curl>
+    $<INSTALL_INTERFACE:include/curl>
+    CACHE PATH "Path to curl.h" FORCE
+  )
+  install(FILES ${CURL_INCLUDE_DIR} DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+endif()
+
+# Get cmake-js::cmake-js via 'cmakejs_create_napi_addon()'
+include(CMakeJS)
+
+cmakejs_create_napi_addon(
+  addon
+  "src/hello/addon.cpp"
+  "src/hello/get.cpp"
+  "src/hello/post.cpp"
+  NAMESPACE vendor
+  EXCEPTIONS YES
+)
+# Move project headers...
+configure_file(
+  "include/hello/addon.h" # from SOURCE_DIR...
+  "include/hello/addon.h" # ..to BINARY_DIR
+)
+target_include_directories(addon PUBLIC
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/hello>
+  $<INSTALL_INTERFACE:include/hello>
+)
```

Comment:
why do this?
whats wrong with including headers from where they naturally reside?

Reply:
This is the really big question that I think is separating our schools of thought. It's the distinction between regular CMake and 'modern' CMake, but not just for the sake of it.

The CMake manual covers this stuff much better than I can, but it's dogged devotion to it's own binary dir does have quite a lot of merits once you scratch the surface.

I will take some time to try to write a more compact roundup on this general subject (the relocatability stuff) because I think we're approaching some decent compromise where we can get this thing around a corner.

Did you happen to notice something though;

The above codeblock is precisely the same change I made to the API that caused some cold feet just a minute ago. Pick up headers and throw them in the binary dir; set the target to include the binary dir path during 'build/config', and to be (silently) prefixed by 'CMAKE_INSTALL_PREFIX' during 'install'.

These are what CMake calls it's 'build' interface and it's 'install' interface.

Classic CMake stops at 'config/build', and cmake-js has worked perfectly by catering to that. In classic CMake, just tell it where the headers are prefixed on your system, and CMake will find it.

Modern CMake allows devs to ship/transport/pack their entire API in a portable way. You saw those export output files earlier, with the ```${_IMPORT_PREFIX}/include/node-addon-api/napi.h``` and everything else? You see how, after running CPack, its' entire output dir moves around, with the headers always sitting at the same relative path to the ones in the Targets.cmake file?

That functionality precisely is what CMake really means by ```install()```. 

Now, I do agree, we had this running great even before I made the targets relocatable. cmake-js has always worked fine without it. So who really needs this?

But please take a good look at the cURL addon project and think; where is cURL coming from? It's not on the package.json, we aren't using vcpkg to get it. If you find your CMake installation dir, you should be able to find a 'Modules' folder in there, with a whole bunch of 'Find*.cmake'. You and I talked about not really getting how those files work - I dont use them much either (I use vcpkg instead). You will have a FindCURL.cmake, which is what your CMake is automatically pulling in for you.

On my Linux driver machine here, the cURL that my CMake file is linked to is really outdated. Nonetheless, you will see in FindCURL.cmake that *yes it does install to /usr/include/, or program files* and yes, that is precisely where both our CMake's and our compilers are finding all of those headers. CMake did not download cURL for us, it just found our installation and used that.

Unfortunately, being that my cURL is old, so is it's CMake config, such that my cURL doesn't have relocatable targets. Rather, I just get a ```CURL::libcurl``` for linkage, and some ```CURL_*``` vars for further playing with.

(Side note - that's why I'm defining stuff like CMAKE_JS_EXE or whatever even if I'm never actually using them in the API; I'm trying to pass those to our users, so that they may also call those vars if they want to do custom stuff with them in their builds. They don't *have* to use these vars. Why not provide this to them? Most CMake API's do export their vars for further usage in this way, and so we align).

So, my old cURL with it's fixed-location headers and vars, does permit my addon to configure and build, just fine. Great! End of story! Well, I can totally see why you think so, but as you can see a bit more clearly now, 'configure/build' is just one part of the story. What if I want to pack my addon, and distribute it?  (I'm sure you're getting that people want to pack and distribute their work obviously)

Well, since cURL had not provided their targets with export/install stuff and did not make them IMPORTED INTERFACE LIBRARY or whatever it is, that means that if I want to do my CMake 'install' stuff on my addon and have cURL as part of the package I distribute, then *I* am the one who has to configure the cURL target to be relocatable, installable, packable, whatever you want to call it. I wouldn't have to do it myself had their CMake been a bit more modernized for my version number.

So what? It's one target, a few lines of code. Yeah, but I'm building these addons like crazy here as you can see. Actually I just dropped 9 working addon projects into the test dirs over the weekend. All of them packable, installable, whatever. If our API did *not* export/install it's targets inside it's CMake file, then I would have had to have put that same code block which is in the API, into each one of these projects one by one, configured each one to get just the right stuff and nothing more (link level, as we're seeing it now)... it would be a major drag, doing that same codeblock 9 times!

Instead, "I" (example user) didn't actually have to write that codeblock at all, because CMakeJS.cmake already did it, did it once and did it correctly. Now that funcionality is there for me to tap into. Doesn't mean I absolutely will make all of my addons installable every time (though personally I do like having the option), but keep this in mind;

By offering this entirely optional functionality, what did our users gain, and what did they lose?

Reply:
I forgot to finish the point about cURL here. My old cURL version forced me to write all the export/install stuff myself in that project, which as you can see is a lot of "stuff", a bit daunting, etc.

More recent version of cURL however, began exporting relocatable targets. This version is the one available on vcpkg. What happens is;

- I start a new C++ project by putting a vcpkg.json at the root
- vcpkg.json is precisely to C++/CMake what npm/yarn is to Javascript
- so my deps all go on my vcpkg.json and I call ```vcpkg install``` in my root dir, and you can guess what happens
- The version of cURL vcpkg provides is definitely not the same as the one that my system is using, so gladly, vcpkg keeps a registry on the disk just like yarn cache, etc
- When I vcpkg install, what happens is that vcpkg unpacks those dep files from disk and copies them into the CMake binary dir (!)
- vcpkg feeds my CMake and gives it a CURL::libcurl, and CMake feeds my IDE, so thanks to vcpkg install, my entire dev env is all hooked up; include paths, link lines, intellisense stuff... "just works"
- vcpkg is fetching, packing, and unpacking it's packages guess how? Yes, CMAKE_INSTALL_PREFIX, the 'install' interface, and of course, CPack
- cURL would not work with vcpkg if it were not relocatable, i.e., if it did not provide an export of it's targets

keep one more thing in mind, if I may;

npm/yarn, mocha (or whatever Js testing framework you like), nodejs, all of that stuff is for *Javascript* packages

vcpkg, CMake, CPack, CTest are tools for *C++* packages.

Which of the two above categories is a Napi Addon? Is it a Javascript project, or is it a C++ project?

I don't know about yourself, but mostly my addons only contain a bindings file in Javascript (maybe 4 lines?) and everything else about it is C++.

Reply:
> It's the distinction between regular CMake and 'modern' CMake

Yeah, this is the key part that I keep forgetting.

It being correct 'modern' cmake is a compelling reason to do this by default for things, but I can't shake that it feels odd to me. I'd be tempted to say lets do it the correct way and see if anyone has any real problems with it. I should try actually making something new using this flow and see if I find any strange issues

> I don't know about yourself, but mostly my addons only contain a bindings file in Javascript (maybe 4 lines?) and everything else about it is C++.

It varies, some have more than others.

For example one has nothing other than the bindings loader, and another is wrapping methods because taking a slice of a buffer is not obvious how to do in c++ but is trivial to do in js.

But again, I mostly do js so unsurprisingly I would naturally drift to js tooling instead of learning ctest :shrug: